explainer.md

```diff
@@ -112,7 +112,36 @@ interface LaunchEvent : ExtendableEvent {
 ```
 Notes:
 * The `Request` that led to the navigation is included. In addition to the URL, this allows sites to inspect things like POST data (e.g. for a Web Share Target) that led to the invocation of the app. 
-## Design Questions/Concerns
+
+## Design Questions/Details
+
+### Restricting to installed apps
+
+It is recommended that user agents only fire launch events for installed web apps. There are 2 reasons for this:
+ 1. It is difficult to attribute bad behavior to misbehaving websites if they aren't installed (see the section below).
+ 2. It could be confusing if the behavior of clicking a link changes just because a user has visited a site that registered itself as a `launch` event handler. 
+
+Allowing launch events to be handled on the drive-by web could be explored in the future.
+
+### Handling Redirects
+
+We should allow launch events to be intercepted when a navigation triggers a redirect. For example imagine that https://www.example.com registers a `launch` event handler. The user then clicks a link to https://example.com. No handler is registered, so the navigation is processed. The navigation redirects to https://www.example.com. At this point, before the request is actually made, a launch event should be fired on the SW for https://www.example.com.
+
+If we don't do this, then many link clicks won't be correctly intercepted.
+
+User agents need to decide what the underlying page will do if a redirect is captured in separate window. The options are:
+- The tab could be closed
+- The tab could continue the navigation in parallel to handling the launch event
+- The tab could stop the partially processed navigation and display an error message
+- The tab could return to the page it was on prior to the navigation starting
+
+### Handling window.open()
+
+Ideally calls to window.open() would also be possible for sites to intercept and handle in a `launch` event handler without the new window opening. One challenge is that it's not sufficient to just examine the URL that's passed into window.open() and intercept it as a launch event as it may trigger a redirect as described above.
+
+As a first cut, we propose not doing anything special for window.open(). A new window will be opened as usual and a navigation started. That navigation can trigger a launch event.
+
+In future we can consider more advanced techniques to avoid opening a new window and improve the user experience.
```

Comment:
I can see this being difficult to do after the fact, if people make sites that depend on these behaviors. 

(i.e. I expect that the new window will get opened so I load content in that window as well as the new app window that gets opened).

Reply:
Interesting thought: I'm pretty sure I could open as many windows as I like with a launch event handler and an out-of-scope redirector:

app installed at malicious.com/app/*
service worker has scope malicious.com/app/*
service worker has a launch event that always causes links to open in a new window.
malicious.com/redirect/?url={URL} 302s to {URL}. This is out of scope.

malicious.com/app/index.html contains the following javascript
```js
const redirectTo = encodeURI('https://malicious.com/index.html');
window.location = `https://malicious.com/redirect?url=${redirectTo}`;
```

`/redirect/` is out of scope, so the out-of-scope to in-scope navigation will trigger a launch event.
`index.html` triggers a new page load (out-of-scope) which redirects in scope, opening a new window.

At the very least, this could cause unlimited windows to open. If we went with the `return to the page prior to navigation solution` (assuming my window.open() only runs once) I could open as many pages on my site as I like. If I wanted my window.open() to run on every load, I would open exponentially more windows.

Should I put this in its own issue?

Reply:
I think this will be resolved by requiring a user gesture and automatically closing a tab/window if it's initial navigation triggers a launch event